Add gemini-2.0-flash-thinking-exp

jlcheng opened the pull request:
Information about this model from the Go SDK for Generative AI:

name: gemini-2.0-flash-thinking-exp
Description: Gemini 2.0 Flash Thinking Experimental
Max tokens: 32767

Additional information from their website:

Thinking Mode is an experimental model and has the following limitations:

32k token input limit
Text and image input only
8k token output limit
Text only output
No built-in tool usage like Search or code execution

As of Dec, 2024, Google says the knowledge cutoff for Gemini-2.0-flash-exp is "August 2024".

jlcheng commented:
I realized that the price is incorrect. Looking at Google AI Studio, the thinking mode is currently free. It is also capped at 10 requests per minute.

karthink commented:
Thanks!  Does the cost metadata need to be updated then?

Note that all Gemini models are currently free, but we mention the costs nevertheless because it applies if you have a paid API account.

kjnez commented:
I tried to add this model in a similar fashion but it seems to be only returning the thinking process, not the generated answers. Here's what I added in `gptel-gemini.el`:
```elisp
(gemini-2.0-flash-thinking-exp-1219
     :description "Gemini 2.0 Flash Thinking Mode is an experimental model that's trained to generate the thinking process the model goes through as part of its response."
     :capabilities (tool json media)
     :mime-types ("image/png" "image/jpeg" "image/webp" "image/heic" "image/heif"
                  "application/pdf" "text/plain" "text/csv" "text/html")
     :context-window 1000000
     ;; input & output price is halved for prompts of 128k tokens or less
     :input-cost 0.0
     :output-cost 0.0
     :cutoff-date "2024-12")
```
Below is what was returned

> *** How to use Emacs use-package's :vc keyword?

> Here's a thinking process to construct the comprehensive explanation about =:vc=:
> 
> 1. *Identify the Core Question:* The user wants to know how to use the =:vc= keyword in =use-package=. This immediately brings up the concept of managing packages not available in standard repositories like MELPA.
> 
> 2. *Establish the Purpose of =:vc=:* The primary reason for =:vc= is to fetch packages directly from version control systems (primarily Git, but also others). This needs to be stated clearly upfront.
> 
> 3. *Identify Key Information Required by =:vc=:*  To fetch a package from VC, you need the repository location and the branch/tag/commit to use. This translates to the need for the repository URL and the =:ref= keyword.
> 
> 4. *Provide a Basic Example:* Start with a simple, common scenario: fetching a package from a Git repository's main branch. This serves as a foundation for more complex cases.
> 
> 5. *Explain the =:ref= Sub-keyword:* Since =:ref= is essential for specifying the version, it deserves its own explanation. Cover common uses like branches, tags, and commits. Include a concrete example for each.
> 
> 6. *Discuss Other Relevant Sub-keywords:*  Consider other keywords that might be used in conjunction with =:vc=. =:recipe= is important for non-standard repository structures or complex installation processes. =:files= is crucial for when you only need specific files.
> 
> 7. *Address Common Scenarios and Use Cases:*
>     * *Development/Bleeding Edge:* Emphasize the benefit of using =:vc= for accessing the latest changes.
>     * *Specific Versions:* Highlight the role of =:ref= in targeting particular releases.
>     * *Local Repositories:* Explain how to use =:vc= with repositories on your local machine.
>     * *Alternative VCS:* Mention the possibility of using other version control systems (though Git is the most common).
> 
> 8. *Provide Best Practices and Considerations:*  Think about advice to make using =:vc= more effective and avoid potential problems:
>     * *Pinning Revisions:*  Emphasize stability by using specific tags or commits.
>     * *Updating Packages:*  Explain how to manually update VC-managed packages.
>     * *Dependency Management:*  Briefly touch upon handling dependencies.
>     * *Security:*  A crucial point regarding running code from potentially untrusted sources.
> 
> 9. *Structure the Explanation Clearly:* Organize the information logically with headings, bullet points, and code examples to enhance readability.
> 
> 10. *Offer a Complete Example:* Combine different aspects of =:vc= into a more comprehensive example that showcases multiple keywords together.
> 
> 11. *Summarize Key Takeaways:*  Reiterate the main points in a concise summary.
> 
> 12. *Review and Refine:*  Read through the explanation to ensure accuracy, clarity, and completeness. Check for any ambiguities or missing information. (For example, initially, I might have forgotten to explicitly mention the different VCS types supported.)  Ensure consistent terminology and formatting.
> 
> *Self-Correction during the process:*
> 
> - *Initial thought:* Focus solely on the basic =:vc= with a Git URL.
> - *Correction:* Realized that =:ref= is equally crucial and needs prominent explanation.
> - *Initial thought:*  Assume everyone uses GitHub.
> - *Correction:* Broadened to mention other Git hosting platforms and the possibility of other VCS (though Git is the primary focus).
> - *Initial thought:*  Not explicitly mention security concerns.
> - *Correction:*  Added a section on security considerations as it's a vital aspect of fetching code from external sources.
> - *Initial thought:*  Provide isolated examples for each sub-keyword.
> - *Correction:* Included a complete example demonstrating the combined use of several keywords.
> 
> By following these steps and engaging in self-correction, the comprehensive and helpful explanation provided in the initial prompt can be constructed.
>

karthink commented:
@kjnez I tried this and based on the log, gptel is printing everything that's sent by the model.  So there's no issue with the parsing.

```


*Response*:
The user wants to understand the ==:vc== keyword in =use-package=. I need to explain its purpose and usage within a =use-package= declaration.This informs =use-package= how to fetch the package if it's not already installed.

Example:

#+begin_src elisp
(use-package some-package
  :vc git
  :recipe (:host github :repo "user/some-package"))
#+end_src

This tells =use-package= to use =git= to clone the package from the specified GitHub repository if the package isn't found.

Reference:  [https://github.com/jwiegley/use-package#package-properties](https://github.com/jwiegley/use-package#package-properties)
```

kjnez commented:
Interesting. I tried the log too and here's what I got:

> *** How to use Emacs use-package's =:vc= keyword?
> 
> The user is asking about the =:vc= keyword in =use-package=. This keyword is used to automatically checkout a package from a version control system if it's not already installed.
> 
> Therefore, the core of the answer should explain that.

karthink commented:
@kjnez By "tried the log", do you mean you looked in the `*gptel-log*` buffer?  It should have a bunch of json.

jlcheng commented:
@karthink 
> Thanks! Does the cost metadata need to be updated then?
> Note that all Gemini models are currently free, but we mention the costs nevertheless because it applies if you have a paid API account.

I think I have a paid API account, but I still can't find pricing info for this model. This is what it says in the AI Studio:

![image](https://github.com/user-attachments/assets/d26b6881-2da8-4038-886e-3a30f8f2b505)

And as far as I can tell by looking in my billing reports, I only have charges for calls to the Gemini 1.5 models. So I guess they're not charging for it?

I removed the pricing info and updated the PR. I manually tested this and was still able to use the model. (And it looks like the code in `gptel--infix-provider` handles missing costs.) LMK if this looks better.

karthink commented:
> LMK if this looks better.

Yup.  Two more changes are needed though:

1. This model supports media (images/documents), so we need to add `:capabilities (media tool)` to the plist.  I'm not sure about tool use but let's add it for now, I have to check this for all the models anyway before merging the tool-use branch.
2. The description is too long -- could you shorten it a bit?  it shows up as an annotation when picking a model via the transient menu, and we'd like it to fit on screen without pushing the other metadata further to the right.  The desired length is about as much as the descriptions for the other Gemini models.

kjnez commented:
> @kjnez By "tried the log", do you mean you looked in the `*gptel-log*` buffer? It should have a bunch of json.

My words are misleading. I mean the committed changes in this pull request. The response I showed above is the one showing up in the main gptel buffer. I seem to be only getting the thinking process each time I try it.

karthink commented:
@kjnez try the following:

1. (setq gptel-log-level 'info)
2. Use this model
3. Check the *gptel-log* buffer.

If what you see in the buffer corresponds to the JSON in the log buffer, there's nothing more gptel can do.

jlcheng commented:
> 2. The description is too long -- could you shorten it a bit? 

Sure. It looks like this now:

![image](https://github.com/user-attachments/assets/0e9d6e98-71ba-4a28-b587-5754f4bd6169)

jlcheng commented:
I had the wrong screenshot in my previous comment and the thinking model was missing in the transient menu.

![image](https://github.com/user-attachments/assets/e5f775bb-13b3-45d7-96ab-3371e6ebb227)

karthink commented:
Yup, this looks fine.

But on an unrelated note, I'm wondering why the fields in the model annotations in Helm aren't aligned for you.   Here's how it appears for me (i) in Vertico and (ii) in the completions buffer without any fancy completions system:

![screenshot_20250103T045851](https://github.com/user-attachments/assets/fe1206f0-023b-4c33-98f3-064eb8019618)

![screenshot_20250103T045938](https://github.com/user-attachments/assets/f768c049-4106-4502-8402-489946f922fa)

karthink commented:
Thanks for the PR @kjnez!

karthink commented:
> My words are misleading. I mean the committed changes in this pull request. The
> response I showed above is the one showing up in the main gptel buffer. I seem
> to be only getting the thinking process each time I try it.

@kjnez there's a bug in the Gemini response parser, I found it.  I'll fix it soon.

karthink commented:
@kjnez I've fixed the issue of missing response text when using the flash-thinking models.

kjnez commented:
Great, thanks.
